### ISSUE TITLE
Tracking Issue for `core::io`

### ISSUE THREAD

**Issue by jmillikin** [rust-lang/rust#117693]
Feature gate: `#![feature(core_io)]`

This is a tracking issue for `core::io`, which contains an OS-independent subset of `std::io`.

### Public API

The initial API of this module consists of [`BorrowedBuf`](https://doc.rust-lang.org/std/io/struct.BorrowedBuf.html) and [`BorrowedCursor`](https://doc.rust-lang.org/std/io/struct.BorrowedCursor.html), which were previously only available in `std`.

### Steps / History

- [ ] Move `BorrowedBuf` and BorrowedCursor` from `std::io`: #...
- [ ] Final comment period (FCP)[^1]
- [ ] Stabilization PR

### Unresolved Questions

- None yet.

[^1]: https://std-dev-guide.rust-lang.org/feature-lifecycle/stabilization.html

**Comment by tgross35** [comment]
These types in core could use a `core::fmt::Write` impl, but I am not sure how this is possible without conflicting with the blanket impl on `std::io::Write`.

**Comment by taralx** [comment]
Is the intent to stabilize this at the same time as #78485 or independently of it?

**Comment by kornelski** [comment]
I'm surprised that this doesn't have any safe integration with `Vec`. I'd like to have `vec.as_borrowed_buf()` that can safely append to the uninitialized capacity, without having to manually perform unsafe `set_len` correction afterwards.

**Comment by keepsimple1** [comment]
Could you please add a link to the PR that originally added `BorrowedBuf` and `BorrowedCursor` ?  (As both this move and the original changes are unstable, it is somewhat difficult to track back the changes).

I believe the link is: https://github.com/rust-lang/rust/pull/97015

**Comment by keepsimple1** [comment]
If I understand, `BorrowedBuf` (like `MaybeUninit`) essentially is safe to write, may I suggest adding a method to the API of `BorrowedBuf`: 
```
/// Copies the bytes of `buf` to self starting at `offset`.
///  Returns the number of bytes actually written.
///
/// If `offset` is equal or larger than `self.buf.len()`, returns 0.
pub fn write(&mut self, offset: usize, buf: &[u8]) -> usize
```

(this method updates `filled` and `init` accordingly) 

The idea is to make the API even easier to use: 
- Simpler: no need to use `BorrowedCursor` for such use cases. 
- No need to use different APIs to write `filled` / `init` section and `uninit` sections.
- No panics (from this method) as it returns the written length.
- Compatible with existing API.

What do you think?

**Comment by hxuhack** [comment]
According to the [source code](https://doc.rust-lang.org/src/core/io/borrowed_buf.rs.html#75), Capacity() returns self.filled, which appears inconsistent with the descriptions of capacity, filled, and initialized.

**Comment by yoshuawuyts** [comment]
.@sunfishcode wrote about an alternative to the `BorrowedBuf` APIs in [Writing into uninitialized buffers in Rust](https://blog.sunfishcode.online/writingintouninitializedbuffersinrust/). The design outlined there is based on the APIs used in [bytecodealliance/rustix](https://github.com/bytecodealliance/rustix), and is based on a new trait `Buffer` that can abstract over both e.g. `Vec<T>` and `&mut [MaybeUninit<T>]`.

While this tracking issue is not the right avenue to discuss the tradeoffs and differences between the two APIs [^zulip], I wanted to make sure that it was mentioned on this tracking issue. It seems like a credible alternative to the `BorrowedBuf` API, and I believe it's worth considering going forward. Or at the very least: the blog post is well worth a read.

[^zulip]: Zulip would be a better avenue for that. I don't believe there is a thread yet, but anyone is free to start one if they want to.

**Comment by briansmith** [comment]
> 1. Zulip would be a better avenue for that. I don't believe there is a thread yet, but anyone is free to start one if they want to. [↩](#user-content-fnref-zulip-2e5cc6e88c7f53063c8bd043b126cb9a)

There are three threads, two of them are old:
* [BorrowedBuf/BorrowedCursor optimized for past, not future](https://rust-lang.zulipchat.com/#narrow/channel/219381-t-libs/topic/BorrowedBuf.2FBorrowedCursor.20optimized.20for.20past.2C.20not.20future)
* [BorrowedBuf/BorrowedCursor would benefit by being generic](https://rust-lang.zulipchat.com/#narrow/channel/219381-t-libs/topic/BorrowedBuf.2FBorrowedCursor.20would.20benefit.20by.20being.20generic)
* [BorrowedCursor::buf lifetimes from BorrowedBuf::unfilled](https://rust-lang.zulipchat.com/#narrow/channel/219381-t-libs/topic/BorrowedCursor.3A.3Abuf.20lifetimes.20from.20BorrowedBuf.3A.3Aunfilled)

**Comment by joshtriplett** [comment]
We've been iterating on this API for a while. I *think* we might be ready to consider stabilization of `BorrowedBuf`/`BorrowedCursor`, and the initial APIs built on them like `read_buf`.

This wouldn't stabilize `borrowed_buf_init`, so that we can potentially remove that later.

Regarding the remaining open question: I've looked closely at the bounds at this point, and we've also gotten comments about potential bounds issues that have now been fixed. I think we've now got the bounds right. For instance, we *do* want always-`Debug`, as printing the entire buffer in `Debug` seems unhelpful.

Regarding other possible designs such as the rustix `Buffer`: I think these are independently useful (even more so now that it's generic), I don't think stabilizing these precludes us from offering something like that in the future, and I also don't think we should delay ecosystem adoption of these on the basis of designing something even more flexible. `Buffer` is awesome, and the ability to easily use a `Vec`'s spare capacity or similar is great, but it has tradeoffs that'll take time to work through, and I'd propose that we do a separate design process for that API.

I'd be happy to write the stabilization PR, once we have consensus on stabilization.

**Comment by joshtriplett** [comment]
Shall we stabilize the `core_io_borrowed_buf` and `read_buf` features?

@rfcbot merge libs-api

**Comment by rust-rfcbot** [comment]
@joshtriplett has proposed to merge this. The next step is review by the rest of the tagged team members:

* [x] @Amanieu
* [ ] @BurntSushi
* [x] @dtolnay
* [x] @joshtriplett
* [ ] @the8472

Concerns:

* ~~must-implement-one-of~~ resolved by https://github.com/rust-lang/rust/issues/117693#issuecomment-5154383915

Once a majority of reviewers approve (and at most 2 approvals are outstanding), this will enter its final comment period. If you spot a major issue that hasn't been raised at any point in this process, please speak up!

See [this document](https://github.com/rust-lang/rfcbot-rs/blob/master/README.md) for info about what commands tagged team members can give me.

**Comment by fintelia** [comment]
Has there been any progress on [this question](https://github.com/rust-lang/rust/issues/117693#issuecomment-2150398589) about better integrating `Vec` and `BorrowedBuf` so that this API can be used on owned buffers without unsafe? Method additions to `Vec` don't necessarily need to land at the same time, but it seems best to at least confirm feasibility before fully stabilizing this feature.

**Comment by joshtriplett** [comment]
@fintelia It seems possible. By way of example, we could have a function on a `Vec` like `with_spare_capacity_buf`, which gives you a buffer to the spare capacity, and afterwards updates the length based on how much of the buffer was filled in.

**Comment by yoshuawuyts** [comment]
> Regarding [rustix Buffer]: I think these are independently useful (even more so now that it's generic), I don't think stabilizing these precludes us from offering something like that in the future […]

Could I ask this point is covered in more detail in any stabilization PR?

The outcome I'd like to avoid here is that stabilizing this API does not preclude later introducing the `Buffer` trait as well. Some more words on how the two are compatible would make for a helpful reference in future discussions.

**Comment by joshtriplett** [comment]
@yoshuawuyts Adding a couple of types, and a couple of functions using those types, does not stop us from later introducing a different trait and functions that use that trait.

**Comment by clarfonthey** [comment]
Note: we'd only stabilise this in `std::io`; `core::io` and `alloc::io` need more time to cook.

**Comment by joshtriplett** [comment]
We discussed this in today's libs-api meeting.

We'd like to rename `BorrowedCursor::written` to `buf_written`, to make it clearer that it includes the amount written to the whole `BorrowedBuf`.

Stabilizing `read_buf` shouldn't happen until we merge https://github.com/rust-lang/rust/pull/106643 (including the use of `rustc_must_implement_one_of`).

Those were the only blockers identified.

**Comment by rust-rfcbot** [comment]
:bell: **This is now entering its final comment period**, as per the [review above](https://github.com/rust-lang/rust/issues/117693#issuecomment-5076437797). :bell:

**Comment by joshtriplett** [comment]
@rfcbot concern must-implement-one-of

**Comment by tgross35** [comment]
Why is FCP happening here rather than https://github.com/rust-lang/rust/issues/78485? As I understand it that is the main issue, this is only tracking use via `core`, which as @clarfonthey mentioned isn't yet possible. That one also has more unresolved questions.

(note that these feature gates have been in need of a cleanup for a while, https://github.com/rust-lang/rust/issues/78485#issuecomment-1965255920)

Regardless, one concern: there are zero examples and some of the API surface isn't even used in std. It would be good to have those added before stabilization in case looking at actual usage makes us think of possible improvements.

**Comment by joshtriplett** [comment]
@rfcbot resolved must-implement-one-of

Addressed in https://github.com/rust-lang/rust/pull/106643 , along with https://github.com/rust-lang/rust/pull/160106 and https://github.com/rust-lang/rust/pull/160250 . See https://github.com/rust-lang/rust/pull/106643#issuecomment-5153499337 for a summary.

**Comment by rust-rfcbot** [comment]
:bell: **This is now entering its final comment period**, as per the [review above](https://github.com/rust-lang/rust/issues/117693#issuecomment-5076437797). :bell:

**Comment by joshtriplett** [comment]
@tgross35 wrote:
> That one also has more unresolved questions.

I reviewed all the unresolved questions in https://github.com/rust-lang/rust/issues/78485#issue-731429529 , and marked them resolved, since all four of them were resolved years ago.

**Comment by joshtriplett** [comment]
> some of the API surface isn't even used in std

As far as I can tell, the only methods not used internally in std are `filled_mut` (but `filled` is used), `into_filled` and `into_filled_mut` (but again, `filled` is used), and `with_unfilled_buf`.

We don't need `filled_mut` and `into_filled_mut` in the standard library because we don't mutate data we've already read; that's a use case that callers outside the standard library have.

We could use `with_unfilled_buf` in places like `Cursor::read_buf` easily enough, to avoid the need for `prev_written`. Feel free to send in a PR switching that over, if it doesn't impact performance.

**Comment by joboet** [comment]
With the generalisation away from `u8` I feel like this is no longer relevant to I/O alone. Is `core::io` really the best place for it, or shouldn't we rather put this in `core::mem` instead?

**Comment by joshtriplett** [comment]
@joboet Interesting idea. It doesn't use anything else in `io`, and it is more about memory buffers. Having it live alongside `MaybeUninit`, which it provides a safer wrapper around, might make sense.

**Comment by programmerjake** [comment]
now that the types are generic, I think it makes more sense for `BorrowedCursor::append` to just require `T: Clone` and not `Copy`. Also, it should probably implement `Extend<T>` or something like that.

**Comment by tgross35** [comment]
> > some of the API surface isn't even used in std
> 
> As far as I can tell, the only methods not used internally in std are `filled_mut` (but `filled` is used), `into_filled` and `into_filled_mut` (but again, `filled` is used), and `with_unfilled_buf`.

Oh I think it's fine that we're not making use of it, I wasn't trying to say std use a blocker. My concern is that there are no examples for how to use this API, and some portions of it (`*filled_mut`, `into_filled*`) are completely uncovered - no documentation example, no tests, no use in std. Pretty simple implementations so I'm not too concerned, but in theory there could be a Miri violation that we have no idea about.

I don't think it's unreasonable to hold off until we have examples since that may also point out any rough edges that are easy to overlook. If somebody is willing to register an examples-and-tests concern on my behalf, I would appreciate it.

**Comment by scottmcm** [comment]
Moving it out of `io` makes a ton of sense to me.  No strong feelings about where -- I'll also mention `core::slice` since this is about dealing with *slices* in a nice way.

I can absolutely imagine things like having a method on `Iterator` to fill a `BorrowedBuf` (which would avoid some problems of `next_chunk` today), and that's something that makes sense even if you're never doing anything `io`-like.

**Comment by scottmcm** [comment]
Concern: `BorrowedCursor::capacity` is surprising.

In all the other `capacity` methods we have, including `BorrowedBuf::capacity`, the capacity is doesn't decrease as you write into the thing.  If you `push` (or `push_within_capacity`, which is a better parallel to writing here) into a `Vec`, the capacity cannot decrease (and for `within_capacity` it stays the same).  But on `BorrowedCursor` if you write to it the capacity decreases.

I think it should get another name.  Maybe something inspired by [`Vec::spare_capacity_mut`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.spare_capacity_mut)?  To emphasize that it decreases as you use it up.

---

I like @tgross35's point that we ought to have more examples for these since seeing how they're used would help catch things like this.

**Comment by tgross35** [comment]
@hanna-kruppe brings up a good point that we have `BorrowedCursor<'data>` and implement convariance with `unsafe`, rather than having `BorrowedCursor<'buf, 'data>` and getting it naturally. This is footgun-prone, there's a lot of `transmute` to launder lifetimes in the implementation.

[#t-libs > Why is it called &#96;io::BorrowedCursor&#96;? @ 💬](https://rust-lang.zulipchat.com/#narrow/channel/219381-t-libs/topic/Why.20is.20it.20called.20.60io.3A.3ABorrowedCursor.60.3F/near/614398036)

**Comment by tgross35** [comment]
Created https://github.com/rust-lang/rust/issues/160476 / https://github.com/rust-lang/rust/pull/160477 for `borrowed_buf_init` which has a separate feature gate so isn't included in the stabilization.

It's probably fine but notable that `advance` is gated under the default `core_io_borrowed_buf` while `advance_checked` is under `borrowed_buf_init`.

**Comment by tgross35** [comment]
Concern: Do we really need both `into_filled` and `into_filled_mut`? We're holding a `&'data mut` anyway, and the following works fine:

```rust
let mut data = [0u8; 10];
let mut x = BorrowedBuf::from(data.as_mut_slice());
let m: &mut [u8] = x.into_filled_mut();
let m1: &[u8] = m;
let m2: &[u8] = m;
dbg!(&m1, &m2);
```

**Comment by tgross35** [comment]
Concern: Should we require `T: Debug` for `impl Debug for BorrowedBuf` and print the buffer? Currently `T` is unconstrained and we just print the length.

**Comment by Ddystopia** [comment]
Concern: `'a` in `BorrowedCursor<'a, T>` is treated as invariant by the borrow checker and will cause problems for embedded developers, who are arguably the target audience. `unsafe` code is used inside to treat it as covariant, but from my manual checks the borrow checker still treats it as invariant. And as far as I am aware there is no way in Rust to declare the variance of a generic argument. Instead, the structure has to be created in a correct way. 

Additionally, in the older version with two lifetimes (`BorrowedCursor<'buf, 'data>`) `'data` is still invariant, which isn't as nice as it could be.

It seems relatively simple to make it covariant, just hold `*const BorrowedBuf<'a, T>` inside.

<details>
<summary> 
I did manual variance checks with this rustc version.
</summary>

```
rustc 1.99.0-nightly (6f72b5dd5 2026-07-22)
binary: rustc
commit-hash: 6f72b5dd5f82226a2773d40efea7bab941892a73
commit-date: 2026-07-22
host: x86_64-unknown-linux-gnu
release: 1.99.0-nightly
LLVM version: 22.1.8
```
</details>

**Comment by tgross35** [comment]
@Ddystopia mind sharing the test code if it's small?

**Comment by Ddystopia** [comment]
> [@Ddystopia](https://github.com/Ddystopia) mind sharing the test code if it's small?

`&'a mut T` is covariant in `'a` and invariant in `T`. `BorrowedCursor<'a, T>`holds `&'a mut BorrowedBuf<'a, T>`, which uses `'a` in both positions: https://doc.rust-lang.org/nomicon/subtyping.html#variance.

```rust
#![feature(core_io_borrowed_buf)]

use std::io::{BorrowedBuf, BorrowedCursor};

// Accepted.
fn buf_covariant<'short, 'long: 'short>(buf: BorrowedBuf<'long, u8>) -> BorrowedBuf<'short, u8> {
    buf
}

// Rejected.
fn cursor_covariant<'short, 'long: 'short>(
    cursor: BorrowedCursor<'long, u8>,
) -> BorrowedCursor<'short, u8> {
    cursor
}

// Rejected.
fn cursor_contravariant<'short, 'long: 'short>(
    cursor: BorrowedCursor<'short, u8>,
) -> BorrowedCursor<'long, u8> {
    cursor
}

fn main() {}
```

And the error (also says that `BorrowedCursor` is invariant over `'a`):

<img width="1156" height="644" alt="Image" src="https://github.com/user-attachments/assets/a5d25ef7-e6c2-454a-a333-9ca6f89ff9dc" />

**Comment by joshtriplett** [comment]
> Concern: Should we require `T: Debug` for `impl Debug for BorrowedBuf` and print the buffer? Currently `T` is unconstrained and we just print the length.

No, because the buffer may be partially uninitialized.

**Comment by tgross35** [comment]
> > Concern: Should we require `T: Debug` for `impl Debug for BorrowedBuf` and print the buffer? Currently `T` is unconstrained and we just print the length.
> 
> No, because the buffer may be partially uninitialized.

Not sure if I'm misunderstanding the objection but I meant we'd print the initialized portion, of course.

**Comment by hanna-kruppe** [comment]
> @hanna-kruppe brings up a good point that we have `BorrowedCursor<'data>` and implement convariance with `unsafe`, rather than having `BorrowedCursor<'buf, 'data>` and getting it naturally. This is footgun-prone, there's a lot of `transmute` to launder lifetimes in the implementation.

I was thrown off by this when I tried to (re-)discover why the API looks the way it does. But if these tricks are sound (I have no idea one way or another) then I think it's reasonable to present a simpler API at cost of an uglier implementation. So no concern about stabilization from my side, for the record. However, I do want to raise two other concerns resulting from that Zulip thread:

Concern 1: The name `BorrowedCursor` has never been clear or intuitive to me. I haven't heard a convincing reason why it's anything like a "cursor", which to me implies some ability to move around relatively freely. @tgross35's idea of `BorrowedWriter` sounds much clearer and accurate to me, as writing to the underlying buffer (sequentially!) is its main/only(?) purpose.

Concern 2: Documentation about how to design APIs using the new types. Per Zulip discussion, it seems that one main reason why `BorrowedCursor` exists at all (as opposed to using `&mut BorrowedBuf` for writing) is because APIs that take `BorrowedCursor` by value are much easier to use correctly than the simpler alternative (with `&mut BorrowedBuf`, it may be swapped out for a `BorrowedBuf` that counts filled elements in an unrelated buffer). This should be clearly spelled out in the docs so people can apply the same to any APIs they design that use the new types, instead of e.g. taking `&mut BorrowedCursor` or even `&mut BorrowedBuf` and doing `buf.unfilled()` internally.

**Comment by joshtriplett** [comment]
> Concern: Do we really need both `into_filled` and `into_filled_mut`? We're holding a `&'data mut` anyway, and the following works fine:
> ```rust
> let mut data = [0u8; 10];
> let mut x = BorrowedBuf::from(data.as_mut_slice());
> let m: &mut [u8] = x.into_filled_mut();
> let m1: &[u8] = m;
> let m2: &[u8] = m;
> dbg!(&m1, &m2);
> ```

Valid. Basing a shared reference on a mutable reference means the original is still mutably borrowed so you can't call methods taking a shared reference, but that's not an issue for an `into` function because you don't have the original to call other methods on anyway.

There's still some value in getting the type you want to begin with, though.

**Comment by fintelia** [comment]
Another annoyance with the current API is that the two `From` implementations for `BorrowedBuf` breaks type inference. Might be better to have dedicated constructors:

```rust
let mut v: Vec<u8> = Vec::with_capacity(10);
let b = BorrowedBuf::from(v.spare_capacity_mut()); // ERROR: cannot infer type for `b`!
```

(The problem is that b could either be a `BorrowedBuf<u8>` or a `BorrowedBuf<MaybeUninit<u8>>`)

**Comment by WaffleLapkin** [comment]
Concern: I don't think the API is ready for stabilization at the moment.

_Numerous_ concerns have been raised: https://github.com/rust-lang/rust/issues/117693#issuecomment-5106705125 (`written` => `buf_written`), https://github.com/rust-lang/rust/issues/117693#issuecomment-5156999054 (moving types out of `io`), https://github.com/rust-lang/rust/issues/117693#issuecomment-5160287875 (relaxing `append` to `T: Clone` from `T: Copy`), https://github.com/rust-lang/rust/issues/117693#issuecomment-5160382935, https://github.com/rust-lang/rust/issues/117693#issuecomment-5170391734 (`capacity` is surprising), https://github.com/rust-lang/rust/issues/117693#issuecomment-5173650638, https://github.com/rust-lang/rust/issues/117693#issuecomment-5173759271, https://github.com/rust-lang/rust/issues/117693#issuecomment-5180050759, https://github.com/rust-lang/rust/issues/117693#issuecomment-5182654686 (2 separate concerns), https://github.com/rust-lang/rust/issues/117693#issuecomment-5183461507, and possibly others I missed.

I think at this point it's clear that the FCP should be canceled, before all the concerns are addressed (even if to some concerns the answer is "we think the current state is fine", we need to explicitly resolve them first) (also note that none of the raised concerns are from libs-api members, so none of them are formally blocking the FCP).

-----

In addition to all said above, I also want to add my own concerns about the API...

Concern 1: The documentation is not in a great state. There are no examples for uses of either of the types, or any of their methods. The documentation text itself is terse and confusing/not helpful in places (especially around the initialized-ness APIs, but also just in general; for example I have no idea what "write-only iterator" is trying to communicate...). Methods are not ordered in a useful order (_the_ important methods are buried under a lot of less important ones)

Concern 2: A few methods unnecessarily return `self`. For example [`clear`](https://doc.rust-lang.org/nightly/alloc/io/struct.BorrowedBuf.html#method.clear) and [`set_init`](https://doc.rust-lang.org/nightly/alloc/io/struct.BorrowedBuf.html#method.set_init) on `BorrowedBuf` or [`advance_checked`](https://doc.rust-lang.org/nightly/alloc/io/struct.BorrowedCursor.html) and [`advance`](https://doc.rust-lang.org/nightly/alloc/io/struct.BorrowedCursor.html#method.advance) on `BorrowedCursor`. This could be useful for chaining methods, but I don't see which methods one would want to chain. _Additionally_ some of these methods are `unsafe`, which makes it even less useful still to chain them (one would probably want to separate the unsafe calls and write safety comments on them).

Concern 3: even though `borrowed_buf_init` is supposedly [not part of the current stabilization](https://github.com/rust-lang/rust/issues/117693#issuecomment-5173597407), it still affects the APIs that _are_ being stabilized. Namely [`as_mut`](https://doc.rust-lang.org/nightly/std/io/struct.BorrowedCursor.html#method.as_mut) has to be `unsafe`, because the caller mustn't write uninitialized values to it, if it's fully initialized[^1]. It's a shame we have to track initialized-ness at all[^2].

Concern 4: `BorrowedBuf` and `BorrowedCursor` methods are gated on `T: Copy` for no reason.

Concern 5: A lot of the names don't feel good. `Read::read_buf` is not a "buffered read", but simply a read into a possibly uninitialized buffer, `BorrowedBuf` and `BorrowedCursor` stress being borrowed and it's not very clear why that is important (isn't it more important here that they are partially uninitialized?), `capacity` is misleading as per https://github.com/rust-lang/rust/issues/117693#issuecomment-5170391734, `advance_checked` can't actually check if `n` elements have been initialized (it only knows if the buffer is fully initialized or not), ...

[^1]: could probably be prevent by always un-initializing the buffer on a call to `as_mut`

[^2]: from my understanding it's only for the purpose of "implementing `read_buf` in terms of `read`", as without tracking it you end up needing to re-initialize the buffer on each call to `read_buf` . still, quite unfortunate, especially for uses outside of `std`, where there might not be a similar concern.

------

In my opinion these APIs need at least a bit more work. And then it would be nicer to stabilize a minimal subset, before the full API.

**Comment by briansmith** [comment]
> Concern 3: even though `borrowed_buf_init` is supposedly [not part of the current stabilization](https://github.com/rust-lang/rust/issues/117693#issuecomment-5173597407), it still affects the APIs that _are_ being stabilized. Namely [`as_mut`](https://doc.rust-lang.org/nightly/std/io/struct.BorrowedCursor.html#method.as_mut) has to be `unsafe`, because the caller mustn't write uninitialized values to it, if it's fully initialized[1](#user-content-fn-1-d657e890fafea9572157f3f9e74584eb). It's a shame we have to track initialized-ness at all[2](#user-content-fn-2-d657e890fafea9572157f3f9e74584eb).

`BorrowedBuf` is roughly `&[MaybeUninit<T>]` except there is no way to write `MaybeUninit::uniniit()` to it without using `unsafe`. `as_mut` has to be `unsafe` to ensure this, and it would have to be `unsafe` regardless of the `borrowed_buf_init` feature.

**Comment by briansmith** [comment]
> Concern 4: `BorrowedBuf` and `BorrowedCursor` methods are gated on `T: Copy` for no reason.

We have `T: Copy` because the API is an abstraction over `[T]` and `[MaybeUninit<T>]` and `T: Copy` is the only way we have to write `T: !Drop` right now. If `T: Drop` then the API becomes dangerous.

**Comment by briansmith** [comment]
> Concern: `'a` in `BorrowedCursor<'a, T>` is treated as invariant by the borrow checker and will cause problems for embedded developers, who are arguably the target audience. 

How does the invariance of `'a` cause problems for embedded developers?

> Additionally, in the older version with two lifetimes (`BorrowedCursor<'buf, 'data>`) `'data` is still invariant, which isn't as nice as it could be.

I such use a two-lifetime variant of `BorrowedCursor` for some embedded-friendly libraries and it hasn't caused me any problems.

> It seems relatively simple to make it covariant, just hold `*const BorrowedBuf<'a, T>` inside.

I saw your PR that implements this proposed solution and it basically disables borrow checking in the implementation of `BorrowedCursor`, making it very difficult to analyze the memory safety of `BorrowCursor`'s implementation.

The main purpose of `BorrowedCursor` is to be a `&'_ mut BorrowedCursor<'a>` to pass to a function `f` that doesn't allow `f` to replace the buffer with a different one.

**Comment by Amanieu** [comment]
We discussed this in the libs meeting, and given the concerns we will cancel this FCP and restart it on a stabilization PR.

@rfcbot cancel

**Comment by rust-rfcbot** [comment]
@Amanieu proposal cancelled.

**Comment by tgross35** [comment]
> We discussed this in the libs meeting, and given the concerns we will cancel this FCP and restart it on a stabilization PR.

There are a lot of things that probably should happen before a stabilization PR. I've collected the issues that need libs-api weigh in, or at least somebody interested enough to do the work, in the unresolved questions to make it easier to see what's done or not.

**Comment by ChrisDenton** [comment]
Should there be an `is_empty` method? Since this is customary when having a `len` method.

**Comment by tgross35** [comment]
> Should there be an `is_empty` method? Since this is customary when having a `len` method.

Added to the checklist.

Nominating since team discussion is needed for at least some of the questions. (A couple are just todo items)

@rustbot label +I-libs-api-nominated, I think this is the right label still?

**Comment by sammhicks** [comment]
I have concerns about [BorrowedCursor::written](https://doc.rust-lang.org/nightly/core/io/struct.BorrowedCursor.html#method.written).

If there's a function which takes a `BorrowedCursor`, you might assume that it's at the start, and thus `written()` returns 0 because you haven't written anything yet. I think that there are going to be subtle logic bugs introduced by this sort of assumption.

As such, I think that instead, `BorrowedCursor` should have a `position` method, which returns an opaque `Position` type. `Position` has relatively few methods, probably just (and their supertraits) `Debug`, `Copy`, `Ord`, and `Sub`. This is analogous to `std::time::Instant`, because to me, you're generally not interested in absolute position, just relative position to some other position, e.g.

```Rust
let start_position = cursor.position();
do_thing(cursor.reborrow());
let end_position = cursor.position();
println!("Wrote {} bytes", end_position - start_position);
```

**Comment by programmerjake** [comment]
`Position` should probably also be `Hash`, and should probably be generic over `T` so it can be a pointer internally (though it'll need a special case for when `T` is size zero).

**Comment by sammhicks** [comment]
> `Position` should probably also be `Hash`, and should probably be generic over `T` so it can be a pointer internally (though it'll need a special case for when `T` is size zero).

I was thinking that it should be usize internally, and constructed with the value of `self.buf.filled` (with self being a BorrowedCursor). Yes, implementing `Hash` seems sensible.
